<a href="https://colab.research.google.com/github/mehreengillani/Data602/blob/main/07_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment 7**
Mehreen Ali Gillani

link: https://colab.research.google.com/drive/1_mJmDGGiG_peh-v3zU4e3BPllByukQP3?usp=sharing
# **Weeks 8 & 9 - Pandas**
* In this homework assignment, you will explore and analyze a public dataset of your choosing. Since this assignment is “open-ended” in nature, you are free to expand upon the requirements below. However, you must meet the minimum requirments as indicated in each section.

* You must use Pandas as the **primary tool** to process your data.

* The preferred method for this analysis is in a .ipynb file. Feel free to use whichever platform of your choosing.  
 * https://www.youtube.com/watch?v=inN8seMm7UI (Getting started with Colab).

* Your data should need some "work", or be considered "dirty".  You must show your skills in data cleaning/wrangling.

### **Some data examples:**
•	https://www.data.gov/

•	https://opendata.cityofnewyork.us/

•	https://datasetsearch.research.google.com/

•	https://archive.ics.uci.edu/ml/index.php

### **Resources:**

•	https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html

•	https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html


### **Headings or comments**
**You are required to make use of comments, or headings for each section.  You must explain what your code is doing, and the results of running your code.**  Act as if you were giving this assignment to your manager - you must include clear and descriptive information for each section.

### **You may work as a group or indivdually on this assignment.**


# Introduction

In this section, please describe the dataset you are using.  Include a link to the source of this data.  You should also provide some explanation on why you choose this dataset.

Dataset Information
Additional Information
This data approach student achievement in secondary education of two Portuguese schools. The data attributes include student grades, demographic, social and school related features) and it was collected by using school reports and questionnaires. Two datasets are provided regarding the performance in two distinct subjects: Mathematics (mat) and Portuguese language (por). In [Cortez and Silva, 2008], the two datasets were modeled under binary/five-level classification and regression tasks. Important note: the target attribute G3 has a strong correlation with attributes G2 and G1. This occurs because G3 is the final year grade (issued at the 3rd period), while G1 and G2 correspond to the 1st and 2nd period grades. It is more difficult to predict G3 without G2 and G1, but such prediction is much more useful (see paper source for more details)

https://archive.ics.uci.edu/dataset/320/student+performance Student Performance Dataset

As a concerned parent, I've always wondered: what factors actually help children do well in school? Are expensive tutors worth it? Does family support really make a difference? Do after-school activities help or hurt grades?

This curiosity led me to analyze a comprehensive student performance dataset that tracks 649 students across various demographics, study habits, family backgrounds, and lifestyle factors. I want to cut through the assumptions and find evidence-based answers about what truly impacts academic success.

______________
# Data Exploration
Import your dataset into your .ipynb, create dataframes, and explore your data.  

Include:

* Summary statistics means, medians, quartiles,
* Missing value information
* Any other relevant information about the dataset.  



In [ ]:
pip install ucimlrepo

In [ ]:
from ucimlrepo import fetch_ucirepo
import pandas as pd
import numpy as np
# fetch dataset
student_performance = fetch_ucirepo(id=320)

# data (as pandas dataframes)
X = student_performance.data.features
y = student_performance.data.targets

# metadata
print(student_performance.metadata)

# variable information
print(student_performance.variables)


{'uci_id': 320, 'name': 'Student Performance', 'repository_url': 'https://archive.ics.uci.edu/dataset/320/student+performance', 'data_url': 'https://archive.ics.uci.edu/static/public/320/data.csv', 'abstract': 'Predict student performance in secondary education (high school). ', 'area': 'Social Science', 'tasks': ['Classification', 'Regression'], 'characteristics': ['Multivariate'], 'num_instances': 649, 'num_features': 30, 'feature_types': ['Integer'], 'demographics': ['Sex', 'Age', 'Other', 'Education Level', 'Occupation'], 'target_col': ['G1', 'G2', 'G3'], 'index_col': None, 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 2008, 'last_updated': 'Fri Jan 05 2024', 'dataset_doi': '10.24432/C5TG7T', 'creators': ['Paulo Cortez'], 'intro_paper': {'ID': 360, 'type': 'NATIVE', 'title': 'Using data mining to predict secondary school student performance', 'authors': 'P. Cortez, A. M. G. Silva', 'venue': 'Proceedings of 5th Annual Future Business Technolo

In [ ]:
# Combine X -features and Y-target, grades in course subject
student_performance_df = pd.concat([X, y], axis=1)

print("Combined DataFrame shape:", student_performance_df.shape)
print("Columns:", student_performance_df.columns.tolist())
print("\nFirst few rows:")
print(student_performance_df.head())

Combined DataFrame shape: (649, 33)
Columns: ['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2', 'G3']

First few rows:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        4   0

In [ ]:
student_performance_df.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='object')

In [ ]:
student_performance_df.describe()

school_name         sex         age     address  family_size  \
count   452.000000  452.000000  452.000000  452.000000   452.000000   
mean      0.252212    0.378319   16.612832    0.258850     0.298673   
std       0.434764    0.485505    1.164179    0.438488     0.458183   
min       0.000000    0.000000   15.000000    0.000000     0.000000   
25%       0.000000    0.000000   16.000000    0.000000     0.000000   
50%       0.000000    0.000000   17.000000    0.000000     0.000000   
75%       1.000000    1.000000   17.000000    1.000000     1.000000   
max       1.000000    1.000000   21.000000    1.000000     1.000000   

       parent_status  mother_education  father_education  travel_time  \
count     452.000000        452.000000        452.000000   452.000000   
mean        0.123894          2.688053          2.469027     1.508850   
std         0.329825          1.133070          1.100863     0.718715   
min         0.000000          0.000000          0.000000     1.000000   
25%         0.000000          2.000000          2.000000     1.000000   
50%         0.000000          3.000000          2.000000     1.000000   
75%         0.000000          4.000000          3.000000     2.000000   
max         1.000000          4.000000          4.000000     4.000000   

       study_time  ...  family_relationships   free_time  workday_alcohol  \
count  452.000000  ...            452.000000  452.000000       452.000000   
mean     2.037611  ...              3.957965    3.097345         1.411504   
std      0.834883  ...              0.896162    1.011811         0.836342   
min      1.000000  ...              1.000000    1.000000         1.000000   
25%      1.000000  ...              4.000000    2.000000         1.000000   
50%      2.000000  ...              4.000000    3.000000         1.000000   
75%      2.000000  ...              5.000000    4.000000         2.000000   
max      4.000000  ...              5.000000    5.000000         5.000000   

       weekend_alcohol      health    absences  grade_period_1  \
count       452.000000  452.000000  452.000000      452.000000   
mean          2.170354    3.495575    3.212389       12.592920   
std           1.229608    1.445991    4.206850        2.240079   
min           1.000000    1.000000    0.000000        0.000000   
25%           1.000000    2.000000    0.000000       11.000000   
50%           2.000000    4.000000    2.000000       12.000000   
75%           3.000000    5.000000    4.000000       14.000000   
max           5.000000    5.000000   32.000000       19.000000   

       grade_period_2  final_grade  average_grade  
count      452.000000   452.000000     452.000000  
mean        12.873894    13.424779      12.963864  
std          2.147453     2.052764       2.039140  
min          9.000000    11.000000       7.333333  
25%         11.000000    12.000000      11.333333  
50%         13.000000    13.000000      12.666667  
75%         14.000000    15.000000      14.333333  
max         19.000000    19.000000      18.666667  

[8 rows x 27 columns]

In [ ]:
student_performance_df.describe(include=object)

mother_job father_job guardian final_grade_category
count         452        452      452                  452
unique          5          5        3                    2
top         OTHER      OTHER   MOTHER                 Good
freq          176        261      310                  321

# Data Wrangling
Create a subset of your original data and perform the following.  

1. Modify multiple column names.

2. Look at the structure of your data – are any variables improperly coded? Such as strings or characters? Convert to correct structure if needed.

3. Fix missing and invalid values in data.

4. Create new columns based on existing columns or calculations.

5. Drop column(s) from your dataset.

6. Drop a row(s) from your dataset.

7. Sort your data based on multiple variables.

8. Filter your data based on some condition.

9. Convert all the string values to upper or lower cases in one column.

10. Check whether numeric values are present in a given column of your dataframe.

11. Group your dataset by one column, and get the mean, min, and max values by group.
  * Groupby()
  * agg() or .apply()

12. Group your dataset by two columns and then sort the aggregated results within the groups.

**You are free (and should) to add on to these questions.  Please clearly indicate in your assignment your answers to these questions.**

In [ ]:
# 1 Modify multiple column names - Rename
student_performance_df = student_performance_df.rename(columns={
    'school': 'school_name',
    'famsize': 'family_size',
    'Mjob': 'mother_job',
    'Fjob': 'father_job',
    'Medu': 'mother_education',
    'Fedu': 'father_education',
    'Pstatus': 'parent_status',
    'traveltime': 'travel_time',
    'studytime': 'study_time',
    'schoolsup': 'school_support',
    'famsup': 'family_support',
    'famrel': 'family_relationships',
    'freetime': 'free_time',
    'goout': 'go_out',
    'reason': 'reason_to_choose_school',
    'Dalc': 'workday_alcohol',
    'Walc': 'weekend_alcohol',
    'G1': 'grade_period_1',
    'G2': 'grade_period_2',
    'G3': 'final_grade'
})

print(student_performance_df.columns)

Index(['school_name', 'sex', 'age', 'address', 'family_size', 'parent_status',
       'mother_education', 'father_education', 'mother_job', 'father_job',
       'reason_to_choose_school', 'guardian', 'travel_time', 'study_time',
       'failures', 'school_support', 'family_support', 'paid', 'activities',
       'nursery', 'higher', 'internet', 'romantic', 'family_relationships',
       'free_time', 'go_out', 'workday_alcohol', 'weekend_alcohol', 'health',
       'absences', 'grade_period_1', 'grade_period_2', 'final_grade'],
      dtype='object')


In [ ]:
#2 Check data types of all columns
print("Data Types:")
print(student_performance_df.dtypes)


Data Types:
school_name                object
sex                        object
age                         int64
address                    object
family_size                object
parent_status              object
mother_education            int64
father_education            int64
mother_job                 object
father_job                 object
reason_to_choose_school    object
guardian                   object
travel_time                 int64
study_time                  int64
failures                    int64
school_support             object
family_support             object
paid                       object
activities                 object
nursery                    object
higher                     object
internet                   object
romantic                   object
family_relationships        int64
free_time                   int64
go_out                      int64
workday_alcohol             int64
weekend_alcohol             int64
health                      int64
ab

In [ ]:
#Based on your data types, I can see several categorical variables that are stored as objects but should be properly encoded.
# First, let's examine the unique values in each object column
object_columns = student_performance_df.select_dtypes(include=['object']).columns

print("Unique values in object columns:")
for col in object_columns:
    print(f"\n{col}: {student_performance_df[col].unique()}")
    print(f"Value counts:")
    print(student_performance_df[col].value_counts())
    print("-" * 40)


Unique values in object columns:

school_name: ['GP' 'MS']
Value counts:
school_name
GP    423
MS    226
Name: count, dtype: int64
----------------------------------------

sex: ['F' 'M']
Value counts:
sex
F    383
M    266
Name: count, dtype: int64
----------------------------------------

address: ['U' 'R']
Value counts:
address
U    452
R    197
Name: count, dtype: int64
----------------------------------------

family_size: ['GT3' 'LE3']
Value counts:
family_size
GT3    457
LE3    192
Name: count, dtype: int64
----------------------------------------

parent_status: ['A' 'T']
Value counts:
parent_status
T    569
A     80
Name: count, dtype: int64
----------------------------------------

mother_job: ['at_home' 'health' 'other' 'services' 'teacher']
Value counts:
mother_job
other       258
services    136
at_home     135
teacher      72
health       48
Name: count, dtype: int64
----------------------------------------

father_job: ['teacher' 'other' 'services' 'health' 'at_home']
Va

In [ ]:
# Convert binary variables (yes/no) to 1/0
binary_columns = ['school_support', 'family_support', 'paid', 'activities',
                  'nursery', 'higher', 'internet', 'romantic']

print("Converting binary variables (yes/no → 1/0):")
for col in binary_columns:
    print(f"{col}: {student_performance_df[col].unique()} → ", end="")
    student_performance_df[col] = student_performance_df[col].map({'yes': 1, 'no': 0})
    print(f"{student_performance_df[col].unique()}")

# Convert binary categorical variables with custom mappings
binary_custom_mappings = {
    'school_name': {'GP': 0, 'MS': 1},        # Gabriel Pereira = 0, Mousinho da Silveira = 1
    'sex': {'F': 0, 'M': 1},                  # Female = 0, Male = 1
    'address': {'U': 0, 'R': 1},              # Urban = 0, Rural = 1
    'family_size': {'GT3': 0, 'LE3': 1},      # Greater than 3 = 0, Less or equal to 3 = 1
    'parent_status': {'T': 0, 'A': 1}         # Living together = 0, Apart = 1
}

print("\nConverting binary categorical variables:")
for col, mapping in binary_custom_mappings.items():
    print(f"{col}: {student_performance_df[col].unique()} → ", end="")
    student_performance_df[col] = student_performance_df[col].map(mapping)
    print(f"{student_performance_df[col].unique()}")

Converting binary variables (yes/no → 1/0):
school_support: ['yes' 'no'] → [1 0]
family_support: ['no' 'yes'] → [0 1]
paid: ['no' 'yes'] → [0 1]
activities: ['no' 'yes'] → [0 1]
nursery: ['yes' 'no'] → [1 0]
higher: ['yes' 'no'] → [1 0]
internet: ['no' 'yes'] → [0 1]
romantic: ['no' 'yes'] → [0 1]

Converting binary categorical variables:
school_name: ['GP' 'MS'] → [0 1]
sex: ['F' 'M'] → [0 1]
address: ['U' 'R'] → [0 1]
family_size: ['GT3' 'LE3'] → [0 1]
parent_status: ['A' 'T'] → [1 0]


Binary (0/1):

school_support, family_support, paid, activities, nursery, higher, internet, romantic: yes=1, no=0
school_name: GP=0, MS=1
sex: F=0, M=1
address: U=0, R=1
family_size: GT3=0, LE3=1
parent_status: T=0, A=1

In [ ]:
# Reset index and make it a column as there is no index
student_performance_df = student_performance_df.reset_index()
student_performance_df = student_performance_df.rename(columns={'index': 'student_id'})

print("DataFrame with reset index:")
student_performance_df.head(2)

DataFrame with reset index:


student_id  school_name  sex  age  address  family_size  parent_status  \
0           0            0    0   18        0            0              1   
1           1            0    0   17        0            0              0   

   mother_education  father_education mother_job  ... family_relationships  \
0                 4                 4    at_home  ...                    4   
1                 1                 1    at_home  ...                    5   

  free_time go_out  workday_alcohol  weekend_alcohol  health  absences  \
0         3      4                1                1       3         4   
1         3      3                1                1       3         2   

   grade_period_1  grade_period_2  final_grade  
0               0              11           11  
1               9              11           11  

[2 rows x 34 columns]

In [ ]:
# Fix missing and invalid values in data.
# Check for missing values
missing_values = student_performance_df.isnull().sum()
print("Missing Values:")
print(missing_values)


Missing Values:
student_id                 0
school_name                0
sex                        0
age                        0
address                    0
family_size                0
parent_status              0
mother_education           0
father_education           0
mother_job                 0
father_job                 0
reason_to_choose_school    0
guardian                   0
travel_time                0
study_time                 0
failures                   0
school_support             0
family_support             0
paid                       0
activities                 0
nursery                    0
higher                     0
internet                   0
romantic                   0
family_relationships       0
free_time                  0
go_out                     0
workday_alcohol            0
weekend_alcohol            0
health                     0
absences                   0
grade_period_1             0
grade_period_2             0
final_grade                

In [ ]:
# Drop duplicate 'student_id' columns if they exist (keeping the first one)
student_performance_df = student_performance_df.loc[:,~student_performance_df.columns.duplicated(keep='first')]

print("UNIQUE VALUES COUNT FOR EACH COLUMN:")
print("=" * 40)
for column in student_performance_df.columns:
    if column != 'student_id':
        unique_count = student_performance_df[column].nunique()
        print(f"{column}: {unique_count} unique values")

print("\n" + "=" * 40)
print("DETAILED VALUE COUNTS:")
print("=" * 40)

for column in student_performance_df.columns:
    if column != 'student_id': # Exclude 'student_id' from value counts as it's a unique identifier
        print(f"\n{column}:")
        print(student_performance_df[column].value_counts())
        print("-" * 30)

UNIQUE VALUES COUNT FOR EACH COLUMN:
school_name: 2 unique values
sex: 2 unique values
age: 8 unique values
address: 2 unique values
family_size: 2 unique values
parent_status: 2 unique values
mother_education: 5 unique values
father_education: 5 unique values
mother_job: 5 unique values
father_job: 5 unique values
reason_to_choose_school: 4 unique values
guardian: 3 unique values
travel_time: 4 unique values
study_time: 4 unique values
failures: 4 unique values
school_support: 2 unique values
family_support: 2 unique values
paid: 2 unique values
activities: 2 unique values
nursery: 2 unique values
higher: 2 unique values
internet: 2 unique values
romantic: 2 unique values
family_relationships: 5 unique values
free_time: 5 unique values
go_out: 5 unique values
workday_alcohol: 5 unique values
weekend_alcohol: 5 unique values
health: 5 unique values
absences: 24 unique values
grade_period_1: 17 unique values
grade_period_2: 16 unique values
final_grade: 17 unique values

DETAILED VALUE 

Data is clean and ready for analysis.

In [ ]:
#4. Create new columns based on existing columns or calculations.
# Average grade across all periods

student_performance_df['average_grade'] = (student_performance_df['grade_period_1'] + student_performance_df['grade_period_2'] + student_performance_df['final_grade']) / 3

In [ ]:
student_performance_df['average_grade'].describe()

count    649.000000
mean      11.625064
std        2.833360
min        1.333333
25%       10.000000
50%       11.666667
75%       13.333333
max       18.666667
Name: average_grade, dtype: float64

In [ ]:
# Grade categories
def grade_category(grade):
    if grade >= 15:
        return 'Excellent'
    elif grade > 10 and grade < 15:
        return 'Good'
    elif grade >= 5 and grade <= 10:
        return 'Satisfactory'
    elif grade < 5:
        return 'Fail'

student_performance_df['final_grade_category'] = student_performance_df['final_grade'].apply(grade_category)

In [ ]:
student_performance_df['final_grade_category'].describe()

count      649
unique       4
top       Good
freq       321
Name: final_grade_category, dtype: object

In [ ]:
#5. Drop column(s) from your dataset.
# List of columns to drop (customize based on your analysis)
columns_to_drop = [
    'student_id',           # it's just an identifier with no predictive power
    'nursery',              # Often less relevant for academic performance
    'go_out',           # go_out with friends
    'romantic',             # May not be relevant for academic analysis
    'reason_to_choose_school' # not relevant for performance prediction
]
# Drop the columns
student_performance_df = student_performance_df.drop(columns=columns_to_drop)

print(f"Remaining columns: {len(student_performance_df.columns)}")
print(f"Dropped {len(columns_to_drop)} columns")

Remaining columns: 31
Dropped 5 columns


In [ ]:
# 6 Drop a row(s) from your dataset.
# Check for duplicates
duplicate_count = student_performance_df.duplicated().sum()
print(f"Found {duplicate_count} duplicate rows")

# Drop duplicates
if duplicate_count > 0:
    student_performance_df = student_performance_df.drop_duplicates()
    print(f"Dropped {duplicate_count} duplicate rows")
    print(f"New shape: {student_performance_df.shape}")



Found 0 duplicate rows


In [ ]:
# Drop rows where final_grade < 10 and store in temp dataframe
temp_failed_students = student_performance_df[student_performance_df['final_grade'] < 11].copy()
student_performance_df = student_performance_df[student_performance_df['final_grade'] >= 11]

print(f"Main dataset: {len(student_performance_df)} students (passed)")
print(f"Temp dataset: {len(temp_failed_students)} students (below average)")

Main dataset: 452 students (passed)
Temp dataset: 197 students (below average)


In [ ]:
# 7. Sort your data based on multiple variables.
# Sort data by multiple variables
student_performance_df = student_performance_df.sort_values(
    by=['final_grade_category', 'final_grade', 'absences'],
    ascending=[True, False, True]
)

print("Data sorted by: final_grade_category (asc), final_grade (desc), absences (asc)")
print(student_performance_df[['final_grade_category', 'final_grade', 'absences']].head(10))

Data sorted by: final_grade_category (asc), final_grade (desc), absences (asc)
    final_grade_category  final_grade  absences
338            Excellent           19         0
636            Excellent           19         4
181            Excellent           18         0
185            Excellent           18         0
332            Excellent           18         0
337            Excellent           18         0
349            Excellent           18         0
509            Excellent           18         0
549            Excellent           18         0
594            Excellent           18         0


In [ ]:
# 8. Filter your data based on some condition.
# final_grade_category = Good & age is greater than 15
student_performance_df[(student_performance_df['final_grade_category'] == 'Good') & (student_performance_df['age'] >15) ]

school_name  sex  age  address  family_size  parent_status  \
22             0    1   16        0            1              0   
73             0    1   16        0            0              0   
190            0    0   16        0            0              0   
236            0    0   16        0            0              0   
247            0    1   16        0            0              0   
..           ...  ...  ...      ...          ...            ...   
369            0    0   19        0            0              0   
44             0    0   16        0            1              0   
79             0    0   16        0            0              0   
118            0    1   17        1            0              0   
405            0    0   19        0            1              1   

     mother_education  father_education mother_job father_job  ... free_time  \
22                  4                 2    teacher      other  ...         5   
73                  3                 1      other      other  ...         3   
190                 2                 2   services      other  ...         4   
236                 2                 3   services    teacher  ...         3   
247                 4                 4     health      other  ...         2   
..                ...               ...        ...        ...  ...       ...   
369                 1                 1      other      other  ...         5   
44                  2                 2      other    at_home  ...         3   
79                  3                 4    at_home      other  ...         4   
118                 1                 3      other      other  ...         2   
405                 2                 3    at_home      other  ...         2   

     workday_alcohol  weekend_alcohol  health  absences  grade_period_1  \
22                 1                3       5         0              12   
73                 2                2       5         0              13   
190                1                4       5         0              13   
236                1                1       3         0              13   
247                2                4       1         0              13   
..               ...              ...     ...       ...             ...   
369                4                3       5        12              10   
44                 2                2       5        14              10   
79                 1                2       3        14              12   
118                1                4       5        14              12   
405                3                4       5        16              10   

     grade_period_2  final_grade  average_grade  final_grade_category  
22               13           14      13.000000                  Good  
73               13           14      13.333333                  Good  
190              12           14      13.000000                  Good  
236              13           14      13.333333                  Good  
247              13           14      13.333333                  Good  
..              ...          ...            ...                   ...  
369              10           11      10.333333                  Good  
44               11           11      10.666667                  Good  
79               11           11      11.333333                  Good  
118              11           11      11.333333                  Good  
405              11           11      10.666667                  Good  

[248 rows x 31 columns]

In [ ]:
# 9. Convert all the string values to upper or lower cases in one column.
# Convert multiple string columns to uppercase
string_columns = ['mother_job', 'father_job', 'guardian']
for col in string_columns:
    student_performance_df[col] = student_performance_df[col].str.upper()

print("All string columns converted to uppercase")
student_performance_df.head(3)

All string columns converted to uppercase


school_name  sex  age  address  family_size  parent_status  \
338            0    0   17        1            1              0   
636            1    1   18        0            0              0   
181            0    0   16        0            0              0   

     mother_education  father_education mother_job father_job  ... free_time  \
338                 3                 1   SERVICES      OTHER  ...         1   
636                 4                 4    TEACHER    TEACHER  ...         2   
181                 4                 2     HEALTH   SERVICES  ...         2   

     workday_alcohol  weekend_alcohol  health  absences  grade_period_1  \
338                1                1       3         0              18   
636                1                4       2         4              17   
181                1                1       3         0              17   

     grade_period_2  final_grade  average_grade  final_grade_category  
338              19           19      18.666667             Excellent  
636              18           19      18.000000             Excellent  
181              17           18      17.333333             Excellent  

[3 rows x 31 columns]

In [ ]:
#10. Check whether numeric values are present in a given column of your dataframe.
# Check if numeric values are present in mother_job column
has_numeric = student_performance_df['mother_job'].str.contains(r'\d', na=False).any()

print(f"Numeric values present in mother_job: {has_numeric}")

Numeric values present in mother_job: False


In [ ]:
# Check any column for numeric values
def check_numeric_in_column(df, column_name):
    has_numeric = df[column_name].astype(str).str.contains(r'\d', na=False).any()
    print(f"Numeric values in {column_name}: {has_numeric}")
    return has_numeric

check_numeric_in_column(student_performance_df, 'age')

Numeric values in age: True


np.True_

In [ ]:
#11. Group your dataset by one column, and get the mean, min, and max values by group. Groupby()
student_performance_df.groupby('final_grade_category').agg({
    'final_grade': ['mean', 'min', 'max'],
    'absences': ['mean', 'min', 'max']
})

final_grade          absences        
                            mean min max      mean min max
final_grade_category                                      
Excellent              16.122137  15  19  2.755725   0  30
Good                   12.323988  11  14  3.398754   0  32

In [ ]:
#11. Group your dataset by one column, and get the mean, min, and max values by group. Groupby()
student_performance_df.groupby('sex').agg({
    'final_grade': ['mean', 'min', 'max'],
    'absences': ['mean', 'min', 'max']
})

final_grade         absences        
           mean min max     mean min max
sex                                     
0     13.594306  11  19  3.16726   0  32
1     13.146199  11  19  3.28655   0  18

In [ ]:
# Group by final_grade_category and get statistics
grade_stats = student_performance_df.groupby('final_grade_category').agg({
    'final_grade': ['mean', 'min', 'max', 'count'],
    'age': ['mean', 'min', 'max']
}).round(2)

print(grade_stats)

                     final_grade                  age        
                            mean min max count   mean min max
final_grade_category                                         
Excellent                  16.12  15  19   131  16.86  15  20
Good                       12.32  11  14   321  16.51  15  21


In [ ]:
 # .apply()
 # Apply function and reset index for cleaner output
result = student_performance_df.groupby('school_name').apply(
    lambda x: x['final_grade'].describe()
).reset_index()

print(result)

final_grade  school_name  count       mean       std   min   25%   50%   75%  \
0                      0  338.0  13.458580  1.996971  11.0  12.0  13.0  15.0   
1                      1  114.0  13.324561  2.216174  11.0  11.0  13.0  15.0   

final_grade   max  
0            19.0  
1            19.0  


/tmp/ipython-input-559291608.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  result = student_performance_df.groupby('school_name').apply(


In [ ]:
#12. # Group by two columns and aggregate
result = student_performance_df.groupby(['school_name', 'family_support']).agg({
    'final_grade': ['mean', 'min', 'max'],
    'absences': ['mean', 'min', 'max']
})

print(result)

                           final_grade          absences        
                                  mean min max      mean min max
school_name family_support                                      
0           0                13.365217  11  18  3.573913   0  18
            1                13.506726  11  19  3.609865   0  32
1           0                13.629630  11  19  2.129630   0  11
            1                13.050000  11  18  2.016667   0  11


# Conclusions  

After exploring your dataset, provide a short summary of what you noticed from this dataset.  What would you explore further with more time?

Key Observations:

1. Data Quality & Structure:

*   Dataset is exceptionally clean with no missing values or duplicates
*   Successfully converted all categorical variables to proper numerical encoding
*   Clear column naming after comprehensive renaming

2. Student Performance Distribution:

*   452 students passed (final_grade ≥ 11), 197 below average (final_grade < 11)
*   Grade categories: 131 "Excellent" (15-19), 321 "Good" (11-14) students
*   Final grades range from 11-19 in the filtered dataset

3. Demographic Insights:

*   Age range: 15-21 years (majority 16-18)
*   Gender distribution: 383 Female (0), 266 Male (1)
*   School distribution: 423 in GP (0), 226 in MS (1)

4. Academic Patterns:

*   Excellent students have slightly higher average age (16.86 vs 16.51)
*   Family support shows mixed impact on grades across schools
*   Absences are relatively low across performance categories













Further Exploration Opportunities:

1. Predictive Modeling:

*   Build models to predict final_grade_category
*   dentify key factors driving academic success

2. School-Specific Analysis:

*   Deep dive into why School 1 shows inverse family support effect
*   Compare teaching methods, resources between schools

3. Temporal Analysis:

*   Track grade progression from period_1 → period_2 → final_grade
*   Identify improvement/decline patterns

4. Family & Environmental Impact:
*   Detailed analysis of parental education/jobs influence
*   Urban vs rural performance differences



